src/Hopper/Parsing/Concrete.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Hopper.Parsing.Concrete where
+
+import Control.Lens (over)
+import qualified Data.Map as Map
+import qualified Data.Text as T
+import qualified Data.Vector as V
+
+import Hopper.Internal.Core.Literal
+import qualified Hopper.Internal.Core.Term as Tm
+import Hopper.Internal.Type.BinderInfo (BinderInfo(BinderInfoData))
+import Hopper.Internal.Type.Relevance (Relevance(Omega))
+import Hopper.Utils.LocallyNameless
+
+data NumericType
+  = IntTy
+  -- | NatTy
+  | RatTy
+  deriving (Show, Eq, Ord, Read)
+
+newtype ConcreteOpId = ConcreteOpId T.Text deriving (Show, Eq, Ord, Read)
+
+data Concrete =
+  Var !T.Text
+  | StrLit !T.Text
+  | NumericLit !(Either Integer Double)
+  -- explicit multiple return values
+  | Return !Concrete
+  | EnterThunk !Concrete
+  | Delay !Concrete
+  | App !Concrete !(V.Vector Concrete)
+  | PrimApp !ConcreteOpId !(V.Vector Concrete)
+  | Lam !(V.Vector T.Text) !Concrete
+  | Let !(V.Vector T.Text)
+        -- | RHS
+        !Concrete
+        -- | Body
+        !Concrete
+  deriving (Show, Eq, Ord, Read)
+
+toPrimOpId :: ConcreteOpId -> PrimOpId
+toPrimOpId (ConcreteOpId opId) = TotalMathOpGmp $ case opId of
+  "+" -> IntAddOpId
+  "+." -> RatAddOpId
+  "-" -> IntSubOpId
+  "-." -> RatSubOpId
+  "*" -> IntMultOpId
+  "*." -> RatMultOpId
+  _ -> error "primop not yet implemented"
+
```

Comment:
channelling the name directed typing forces of SML and OCAML you are. 
though I think we'll all be happier once things are more typeclass esque / type directed names.

I am a bit confused about the presumption here that all ops are infix.... is that really sensible?